VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,989 @@
+  a package management tools will determine and select the set of package
+  versions that satisfy all the version ranges constraints of all dependencies.
+  This usually requires deploying heuristics and algorithms (possibly complex
+  such as sat solvers) that are ecosystem- and tool-specific and outside of the
+  scope for this specification; yet ``vers`` could be used in tandem with
+  ``purl`` to provide an input to this dependencies resolution process.
+
+
+Examples
+~~~~~~~~~
+
+A single version in an npm package dependency:
+
+- originally seen as a dependency on version "1.2.3" in a package.json manifest
+- the version range spec is: ``vers:npm/1.2.3``
+
+
+A list of versions, enumerated:
+
+- ``vers:pypi/0.0.0|0.0.1|0.0.2|0.0.3|1.0|2.0pre1``
+
+
+A complex statement about a vulnerability in a "maven" package that affects
+multiple branches each with their own fixed versions at 
+https://repo1.maven.org/maven2/org/apache/tomee/apache-tomee/ 
+Note how the constraints are sorted:
+
+
+- "affects Apache TomEE 8.0.0-M1 - 8.0.1, Apache TomEE 7.1.0 - 7.1.2,
+  Apache TomEE 7.0.0-M1 - 7.0.7, Apache TomEE 1.0.0 - 1.7.5."
+
+- a normalized version range spec is:
+  ``vers:tomee/>=1.0.0-beta1|<=1.7.5|>=7.0.0-M1|<=7.0.7|>=7.1.0|<=7.1.2|>=8.0.0-M1|<=8.0.1``
+
+- alternatively, four ``vers`` express the same range, using one ``vers`` for
+  each vulnerable "branches": 
+  - ``vers:tomee/>=1.0.0-beta1|<=1.7.5``
+  - ``vers:tomee/>=7.0.0-M1|<=7.0.7``
+  - ``vers:tomee/>=7.1.0|<=7.1.2``
+  - ``vers:tomee/>=8.0.0-M1|<=8.0.1``
+
+Conversing Rubygems custom syntax for dependency on gem. Note how the
+pessimistic version constraint is expanded:
+
+- ``'library', '~> 2.2.0', '!= 2.2.1'``
+- the version range spec is: ``vers:gem/>=2.2.0|!= 2.2.1!<2.3.0``
+
+
+URI scheme
+~~~~~~~~~~
+
+The ``vers`` URI scheme is  an acronym for "VErsion Range Specifier".
+It has been selected because it is short, obviously about version and available
+for a future formal registration for this URI-scheme at the IANA registry.
+
+The URI scheme is followed by a colon ":".
+
+
+``<versioning-scheme>``
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``<versioning-scheme>`` (such as ``npm``, ``deb``, etc.) determines:
+
+- the specific notation and conventions used for a version string encoded in
+  this scheme. Versioning schemes often specify a version segments separator and
+  the meaning of each version segments, such as [major.minor.patch] in semver.
+
+- how two versions are compared as greater or lesser to determine if a version
+  is within or outside a range.
+
+- how a versioning scheme-specific range notation can be transformed in the
+  ``vers`` simplified notation defined here.
+
+By convention the versioning scheme **should** be the same as the ``Package URL``
+package type for a given package ecosystem. It is OK to have other schemes
+beyond the purl type. A scheme could be specific to a single package name.
+
+The ``<versioning-scheme>`` is followed by a slash "/".
```

Comment:
> Maybe it is a good idea to define a "private space" for non / not yet official versioning schemes? This would avoid conflicting use.

Could you elaborate? What would this look like?

Reply:
BCP 47 uses (different) specific ranges. HTTP Header reserved the `X-` space (and later deprecated that because of the inconveniences it caused when non-standard fields became standard). Obviously, there are things to consider here:
1. How many version schemes could exist? (Probably as many as developers producing software...)
2. How often / how fast can a new scheme be listed? (If that is very fast, the private space is not a problem...)
3. How many software will support purl (or be described by purl) but not follow one of the listed schemes? (As it is proposed to be used in security advisories: a lot.)

My request would be: Provide (not necessarily in the standard) guidance on:
- how new schemes can be added, 
- what should be done, if the scheme does not yet exist,
- how the single developer project can be handled who wants to use vers but probably won't be able to register it's own scheme

Reply:
BTW: Will the registry be handled by IANA or through purl?

Reply:
> BTW: Will the registry be handled by IANA or through purl?

I can see the `vers` URI scheme registered at IANA and versioning schemes registered here through PR otherwise.

Reply:
Well - I need to think again what the suggestion would look like (considering the experience other standards had in the past...). I like the "easiness" of a PR but IANA would be my preferred option...

Reply:
Having talked about this, I think we need 2 things:
- Describe how need version scheme can be added (you mention that this needs a spec; I would suggest: spec, testcases, and reference implementation). The exact process can be outside of this spec (so that we don't need to change the spec if the process changes - makes it easier for people to point to a specific version of the spec).
- Add a reserved private namespace (see https://github.com/package-url/purl-spec/issues/268)